File: clone_graph.py
class UndirectedGraphNode:
    def __init__(self,x):
        self.label =x
        self.neighbors=[]
class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node is None:
            return None
        return self.dfs(node,{})
    def dfs(self,node,dic):
        if node in dic:
            return dic[node]
        out=UndirectedGraphNode(node.label)
        dic[node]=out
        for neighbor in node.neighbors:
            out.neighbors.append(self.dfs(neighbor,dic))
        return out
if __name__=='__main__':
    sol=Solution()
    
    print sol.cloneGraph(node)

File: Algorithm/CPP/clone_graph.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <algorithm>
using namespace std;


class Solution {
public:
	unordered_map<UndirectedGraphNode*, UndirectedGraphNode*>myMap;
	UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
	
		if (!node) return node;
		return dfs(node);
	}

	UndirectedGraphNode* dfs(UndirectedGraphNode* node)
		{
			if (myMap.find(node) != myMap.end())
				return myMap[node];
			UndirectedGraphNode* root = new UndirectedGraphNode(node->label);
			myMap[node] = root;
			for (auto neightbor : node->neighbors)
			{
				root->neighbors.push_back(	dfs(neightbor));
			}
		return root;
	}
};

